packages/firestore/src/util/async_queue.ts

```diff
@@ -279,6 +316,41 @@ export class AsyncQueue {
     return this.enqueueInternal(op);
   }
 
+  enqueueRetryable(op: () => Promise<void>): void {
+    this.verifyNotFailed();
+
+    if (this._isShuttingDown) {
+      return;
+    }
+
+    if (this.firstRetryableOperation) {
+      // If there is already a retryable operation, enqueue the current
+      // operation right after. We want to successfully run the first retryable
+      // operation before running all others.
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.firstRetryableOperation.then(() => {
+        this.enqueueRetryable(op);
+      });
+    } else {
+      const deferred = new Deferred<void>();
+      this.firstRetryableOperation = deferred.promise;
+
+      const retryingOp = async (): Promise<void> => {
+        try {
+          await op();
+          deferred.resolve();
+          this.backoff.reset();
+          this.firstRetryableOperation = undefined;
+        } catch (e) {
+          logDebug(LOG_TAG, 'Retryable operation failed: ' + e.message);
+          this.backoff.backoffAndRun(retryingOp);
+        }
+      };
+
+      this.backoff.backoffAndRun(retryingOp);
```

Comment:
The streams try to perform their initial operation and only call `backoffAndRun` if there's been an initial failure. I wonder if your testing difficulties could be resolved if you called `this.enqueueAndForget` or similar here for the first iteration? That way you'd only be going through the backoff after a failure.

Reply:
We made it `backoffAndRun` rather than `runWithBackoff` specifically because we were intending that the first invocation is after a failure, and the action taken is to backoff then run.

The backoff delay of zero for the first iteration is intentional, at least in the stream case. We found that e.g. connections could get in a stale state or watch could return spurious errors after a connection or database had been idle for a while.

TransactionRunner seems like it's using the API wrong, at least according to the design intent. We have the same problem there where connections can be stale and fail so it's worth avoiding delay on the first failure, even for transactions.

In the case of IndexedDB failures, the initial zero delay doesn't seem as useful, but it also doesn't seem harmful. If you wanted to make the zero delay on first retry optional that would be fine with me too.

Reply:
I think for IndexedDB failures, if would be "feel better" if we retried with backoff right away.  We already retry aborted transactions immediately, since these are likely to succeed even without backoff. Thus, I would prefer if the secondary error-code unaware retry mechanism was different and used some sort of backoff. 
It does seem like this unnecessarily complicates our tests though. I would suggest we defer this behavior change until we have actual evidence that backing off is an improvement.